File: 6pset/Credit/credit.py
# Checks validity of a credit card according to Luhn’s algorithm
from cs50 import get_int
import sys


def main():
    while True:
        # Get credit card number
        creditCard = get_int("Number: ")
        
        # Analyze user input number
        if creditCard > 0:
            total = 0
            card_num = str(creditCard)
            l = len(card_num)
            
            # Curate desired creditcard number length
            if l not in [13, 15, 16]:
                print("INVALID")
                sys.exit(0) 
            
            # Reverse card number
            rev_Cardnum = card_num[::-1]
            
            # Luhn’s Algorithm calculation   
            for n in rev_Cardnum[0::2]:
                total += int(n)
            
            # Calculating from second last number
            for c in rev_Cardnum[1::2]:
                x = str(int(c) * 2)
                
                if x[0] == '1':
                    total += (1 + int(x[1]))
                else:
                    total += int(x)
            
            # Validate credit card number
            if total % 10 != 0:         
                print("INVALID") 
                sys.exit(0)
        
            # Print card type    
            if int(card_num[:2]) in [51, 52, 53, 54, 55] and l == 16:
                print("MASTERCARD")
            elif int(card_num[:2]) in [34, 37] and l == 15:
                print("AMEX")
            elif int(card_num[:1]) == 4 and l in [13, 16]:
                print("VISA") 
                
            sys.exit(0)
        
        
if __name__ == "__main__":
    main()    

File: 1pset/credit/credit.c
#include <stdio.h>
#include <cs50.h>



int main(void)
{
    // Get credit card number
    long creditNum;
    do
    {
        creditNum = get_long("Number: ");
    }
    while (creditNum < 0);
    
    
    int c = 0; 
    int even = 0; 
    int odd = 0;
    int sum = 0;
    long number = creditNum;
    
    // First step of Luhn's Algorithm
    while (creditNum)
    {
        int n = 0; 
        int r = 0;
        c++;
        
        // Gets last digit of creditNum
        n = creditNum % 10;  
        
        // Selects every other digit starting with 
        // the second-to-last-digit
        if (!(c % 2)) 
        {
            n *= 2;
            if (n >= 10)
            {
                // Adds each digit of n to the overall sum
                even += 1; 
                r = n % 10;
                even += r;
            }
            else
            {
                even += n;
            }
        }
        else
        {
            odd += n;
        }
        
        // removes the last digit from creditNum
        creditNum /= 10; 
    }
    
    // Second step of Luhn's Algorithm 
    sum = odd + even;
    
    // Third step of Luhn's Algorithm (validating)
    if (!(sum % 10))
    {
        int count = 0;
        bool MASTERCARD = false;
        bool AMEX = false;
        int n;
        
        while (number)
        {
            count++;
            n = number % 10;
            number /= 10;
            
            // Checks for the 2nd digit of 
            // Mastercard and AMEX 
            if (count == (c - 1))
            {
                // If 2nd digit is present set card to true
                switch (n)
                {
                    case 1 :
                        MASTERCARD = true;
                        break;
                    case 2 :
                        MASTERCARD = true;
                        break;
                    case 3 :
                        MASTERCARD = true;
                        break;
                    case 4 :
                        MASTERCARD = true;
                        AMEX = true;
                        break;
                    case 5 :
                        MASTERCARD = true;
                        break;
                    case 7 :
                        AMEX = true;
                        break;
                }
            }
            
            // Checks the first digit n of the card number accordingly  
            // Checks for the length of the card number accordingly
            if (count == c)
            {
                switch (n)
                {
                    case 3 : 
                        if (AMEX && (c == 15))
                        {
                            printf("AMEX\n");
                        }
                        else
                        {
                            printf("INVALID\n");
                        }
                        break;
                    case 4 :
                        if (c == 13 || c == 16)
                        {
                            printf("VISA\n");
                        }
                        else
                        {
                            printf("INVALID\n");
                        }
                        break;
                    case 5 :
                        if (MASTERCARD && (c == 16))
                        {
                            printf("MASTERCARD\n");
                        }
                        else
                        {
                            printf("INVALID\n");
                        }
                        break;
                    default :
                        printf("INVALID\n");    
                }
            }
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
